c10/core/impl/DeviceGuardImplInterface.h

```diff
@@ -360,6 +360,18 @@ inline const DeviceGuardImplInterface* getDeviceGuardImpl(DeviceType type) {
   static_assert(sizeof(DeviceType) == 1, "DeviceType is not 8-bit");
   auto p = device_guard_impl_registry[static_cast<size_t>(type) & 0xFF].load();
 
+  if (type == DeviceType::CUDA) {
+    if (p == nullptr) {
+      // Trying to get a cuda device guard on a cpu-only build. Return a
+      // no-op device guard.
+      return new NoOpDeviceGuardImpl<DeviceType::CUDA>();
+    } else if (p != nullptr && p->deviceCount() == 0) {
+      // cuda build enabled, but no cuda devices available. Return a
+      // no-op device guard.
+      return new NoOpDeviceGuardImpl<DeviceType::CUDA>();
+    }
+  }
+
```

Comment:
I'm fine with how this looks. Can you add a test?

Reply:
test added. 

notice that I also change the impl to override the guard registry.